src/lib/snark_work_lib/partitioned.ml

```diff
@@ -0,0 +1,534 @@
+open Core_kernel
+
+module Pairing = struct
+  module ID = struct
+    [%%versioned
+    module Stable = struct
+      module V1 = struct
+        (* this identifies a One_or_two work from Work_selector's perspective *)
+        type t = Pairing_ID of int
+        [@@deriving compare, hash, sexp, yojson, equal]
+
+        let to_latest = Fn.id
+      end
+    end]
+  end
+
+  (* A Pairing.Single.t identifies one part of a One_or_two work *)
+  module Single = struct
+    [%%versioned
+    module Stable = struct
+      module V1 = struct
+        (* Case `One` indicate no need to pair. *)
+        type t = [ `First of ID.Stable.V1.t | `Second of ID.Stable.V1.t | `One ]
+        [@@deriving compare, hash, sexp, yojson, equal]
+
+        let to_latest = Fn.id
+      end
+    end]
+  end
+
+  (* A Pairing.Sub_zkapp.t identifies a sub-zkapp level work *)
+  module Sub_zkapp = struct
+    [%%versioned
+    module Stable = struct
+      module V1 = struct
+        (* Case `One` indicate no need to pair. ID is still needed because zkapp command
+           might be left in pool of half completion. *)
+        type t =
+          { which_one : [ `First | `Second | `One ]; id : ID.Stable.V1.t }
+        [@@deriving compare, hash, sexp, yojson, equal]
+
+        let to_latest = Fn.id
+      end
+    end]
+
+    let of_single (id_gen : unit -> ID.t) : Single.t -> t = function
+      | `First id ->
+          { which_one = `First; id }
+      | `Second id ->
+          { which_one = `Second; id }
+      | `One ->
+          { which_one = `One; id = id_gen () }
+  end
+end
+
+module Spec_common = struct
+  [%%versioned
+  module Stable = struct
+    module V1 = struct
+      type t =
+        { (* When is this spec issued? *)
+          issued_since_unix_epoch : Mina_stdlib.Time.Span.Stable.V3.t
+              (* The fee of the full command, even if we're issuing a sub_zkapp level spec *)
+        ; fee_of_full : Currency.Fee.Stable.V1.t
+        }
+      [@@deriving sexp, yojson]
+
+      let to_latest = Fn.id
+    end
+  end]
+end
+
+module Zkapp_command_job = struct
+  (* This identifies a single `Zkapp_command_job.t` *)
+  module ID = struct
+    [%%versioned
+    module Stable = struct
+      module V1 = struct
+        type t = Job_ID of int [@@deriving compare, hash, sexp, yojson]
+
+        let to_latest = Fn.id
+      end
+    end]
+  end
+
+  module Spec = struct
+    [%%versioned
+    module Stable = struct
+      [@@@no_toplevel_latest_type]
+
+      module V1 = struct
+        type t =
+          | Segment of
```

Comment:
IIUC, it is because either we merge two proofs, or either we are generating one.

Reply:
Yes indeed